crates/holochain_zome_types/src/countersigning.rs

```diff
@@ -0,0 +1,288 @@
+    entry_hash: EntryHash,
+}
+
+/// All the data required for a countersigning session.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CounterSigningSessionData {
+    preflight_request: PreflightRequest,
+    responses: Vec<CounterSigningAgentState>,
+}
+
+/// Build an unsigned Create header from session data, shared create data and an agent's state.
+impl
+    From<(
+        &CounterSigningSessionData,
+        &CreateBase,
+        &CounterSigningAgentState,
+    )> for Create
+{
+    fn from(
+        (session_data, create_base, agent_state): (
+            &CounterSigningSessionData,
+            &CreateBase,
+            &CounterSigningAgentState,
+        ),
+    ) -> Self {
+        Create {
+            author: session_data.preflight_request.signing_agents[agent_state.agent_index as usize]
+                .0
+                .clone(),
+            timestamp: Timestamp(
+                session_data
+                    .preflight_request
+                    .session_times
+                    .start
+                    .0
+                    .checked_add(SESSION_HEADER_TIME_OFFSET_MILLIS)
+                    .unwrap_or(i64::MAX),
+                session_data.preflight_request.session_times.start.1,
+            ),
+            header_seq: agent_state.header_seq,
+            prev_header: agent_state.chain_top.clone(),
+            entry_type: create_base.entry_type.clone(),
+            entry_hash: create_base.entry_hash.clone(),
+        }
+    }
+}
+
+/// Build an unsigned Update header from session data, shared update data and an agent's state.
+impl
+    From<(
+        &CounterSigningSessionData,
+        &UpdateBase,
+        &CounterSigningAgentState,
+    )> for Update
+{
+    fn from(
+        (session_data, update_base, agent_state): (
+            &CounterSigningSessionData,
+            &UpdateBase,
+            &CounterSigningAgentState,
+        ),
+    ) -> Self {
+        Update {
+            author: session_data.preflight_request.signing_agents[agent_state.agent_index as usize]
+                .0
+                .clone(),
+            timestamp: Timestamp(
+                session_data
+                    .preflight_request
+                    .session_times
+                    .start
+                    .0
+                    .checked_add(SESSION_HEADER_TIME_OFFSET_MILLIS)
+                    .unwrap_or(i64::MAX),
+                session_data.preflight_request.session_times.start.1,
+            ),
+            header_seq: agent_state.header_seq,
+            prev_header: agent_state.chain_top.clone(),
+            original_header_address: update_base.original_header_address.clone(),
+            original_entry_address: update_base.original_entry_address.clone(),
+            entry_type: update_base.entry_type.clone(),
+            entry_hash: update_base.entry_hash.clone(),
+        }
+    }
+}
+
+impl From<CounterSigningSessionData> for Vec<Header> {
```

Comment:
I also don't see why this can't just be a simple method on `CounterSigningSessionData`.